seperate dirmonitor logic, add build time detection of features

Jan200101 opened the pull request:
instead of hoping that Linux has inotify and everything else has kqueue we detect those features at build and give the option to choose one at build time.

The logic was seperated into
- dirmonitor_inotify.c
- dirmonitor_kqueue.c
- dirmonitor_win32.c

Jan200101 commented:
fixed windows dirmonitor logic

moved dirmonitor files into their own directory as suggested by guldo

adamharrison commented:
I think this is definitely the way to go; having a build option to set the logic internally.

That said, I'm not super keen on splitting up the file into a header and three sub-files; it's a different style from the rest of the APIs (all of which define their functions statically with no header); and there's enough overlapping logic in the functions, that I think it does make sense to keep them together (though this is just my personal stylistic opinion). 

This, however, also does introduce an element of conditional compilation of files depending on options, whereas before, we just compile everything with *.c indiscriminately.

I may be slightly biased here, because I don't build with meson, but I'd argue that if possible, we should just keep it as simple as possible. (i.e. we just compile everything unconditionally, and let the C preprocessor do the work).

Also, IMHO, ideally we could simply set a default define if the user doesn't define one (i.e. #ifndef DIRMONITORBACKEND, #define one based on OS; rather than just #error); that'll work for 99% of people just out of the box.

jgmdev commented:
> but I'd argue that if possible, we should just keep it as simple as possible. (i.e. we just compile everything unconditionally, and let the C preprocessor do the work).

That should be possible by adding to `src/api/dirmonitor.c` something like:

```c
#ifdef _WIN32
  #include "dirmonitor/dirmonitor_win32.c"
#elif __linux__
  #include "dirmonitor/dirmonitor_inotify.c"
#else
  #include "dirmonitor/dirmonitor_kqueue.c"
#endif
```

adamharrison commented:
> That should be possible by adding to src/api/dirmonitor.c something like:

But then we're not unconditionally compiling everything, and it's arguably more confusing, because now you're only supposed to compile that file, and not the sub-files (i.e. you're omitting all backends from compilation).

jgmdev commented:
> But then we're not unconditionally compiling everything, and it's arguably more confusing, because now you're only supposed to compile that file, and not the sub-files (i.e. you're omitting all backends from compilation).

I see, well one could use some define flags instead so it is easier to build with alternate build systems like make eg:

```c
#ifdef DIRMON_WIN32
  #include "dirmonitor/dirmonitor_win32.c"
#elif DIRMON_INOTIFY
  #include "dirmonitor/dirmonitor_inotify.c"
#else
  #include "dirmonitor/dirmonitor_kqueue.c"
#endif
```

adamharrison commented:
Definitely, if we go that route, that's what we should do.

I'm just saying, I feel like it was simpler if it's all just a single file, like the other APIs.

jgmdev commented:
taking a look at the code all you would need is to define `-DDIRMONITOR_BACKEND=backend`

adamharrison commented:
Yeah; I understand that, I'm saying from a *looking at the file and groking it* point of view. Like, as it stands, if you look at one API, they're all the same. A single C file that defines static functions, with one "open" function that gets called in api.c. All files that end in *.c are compiled.

We're changing this to have conditional compilation of the .c files, now with a header just for this API, that will now error out if you *don't* define something.

I do think we should be able to specify the back-end with a define; no question; it's a great idea. I just think that it'd be simpler if instead of changing those things, we just kept it in line with the other APIs, and the old compilation assumptions.

Jan200101 commented:
Okay lets get into this

> I think this is definitely the way to go; having a build option to set the logic internally.

agreed

> That said, I'm not super keen on splitting up the file into a header and three sub-files; it's a different style from the rest of the APIs (all of which define their functions statically with no header); and there's enough overlapping logic in the functions, that I think it does make sense to keep them together (though this is just my personal stylistic opinion).

this is practically the best way to do it
sadly a lot of the dirmonitor code is dependent on the struct and there is no simple way of telling it it exists without informing it of its structure.

The header also allows us to cheat by setting defining the functions for a specific backend using preprocessors, which means any explit use of a certain backend will instantly cause compilation errors as well as define the function currectly so the compiler will not make assumptions

> This, however, also does introduce an element of conditional compilation of files depending on options, whereas before, we just compile everything with *.c indiscriminately.

> I may be slightly biased here, because I don't build with meson, but I'd argue that if possible, we should just keep it as simple as possible. (i.e. we just compile everything unconditionally, and let the C preprocessor do the work).

There is no easy way to do this with just preprocessors.
We can't detect what functions we have or what headers we can include and, at best, can guess.
There is the solution of defining `kqueue` and `inotify_init` in our code and dynamically resolving those at runtime, but I think that is way too overkill

> ~Also, IMHO, ideally we could simply set a default define if the user doesn't define one (i.e. #ifndef DIRMONITORBACKEND, #define one based on OS; rather than just #error); that'll work for 99% of people just out of the box.~ EDIT: Actually, this would be a semi-duplication of logic, so not ideal. I'm OK with it for now, I suppose.

Currently in `src/meson.build` we detect the if the user is capable of using inotify, kqueue (either through native kqueue or libkqueue) or if they are on Windows, in which case we use the win32 backend.

In theory the user should never be able to set an empty DIRMONITOR_BACKEND with our build system, this is just to prevent errors for people that compile them with their own script or similar.

> EDIT2: Although; actually, now that I'm thinking about it, do we want to support the case where _no_ backend is available, instead of erroring out? The current system would support just falling back to scanning in that case.

The current system would not compile either unless at least one backend is present.
It is trivial to add support for no backend or add a generic posix backend.



> > but I'd argue that if possible, we should just keep it as simple as possible. (i.e. we just compile everything unconditionally, and let the C preprocessor do the work).
> 
> That should be possible by adding to `src/api/dirmonitor.c` something like:
> 
> ```c
> #ifdef _WIN32
>   #include "dirmonitor/dirmonitor_win32.c"
> #elif __linux__
>   #include "dirmonitor/dirmonitor_inotify.c"
> #else
>   #include "dirmonitor/dirmonitor_kqueue.c"
> #endif
> ```

I am very certain that this will mess up with build caching since it won't recompile dirmonitor.c unless it changed.



> > But then we're not unconditionally compiling everything, and it's arguably more confusing, because now you're only supposed to compile that file, and not the sub-files (i.e. you're omitting all backends from compilation).
> 
> I see, well one could use some define flags instead so it is easier to build with alternate build systems like make eg:
> 
> ```c
> #ifdef DIRMON_WIN32
>   #include "dirmonitor/dirmonitor_win32.c"
> #elif DIRMON_INOTIFY
>   #include "dirmonitor/dirmonitor_inotify.c"
> #else
>   #include "dirmonitor/dirmonitor_kqueue.c"
> #endif
> ```

I am more fond of the DIRMONTOR_<BACKEND> build option solution, and it works well for projects like SDL2.
Problem I see is that with meson we cannot programatically define options, so unless we are willing to let the meson_options.txt bloat I see this as unviable (also there is no #elif DIRMONITOR_KQUEUE`




> Yeah; I understand that, I'm saying from a _looking at the file and groking it_ point of view. Like, as it stands, if you look at one API, they're all the same. A single C file that defines static functions, with one "open" function that gets called in api.c. All files that end in *.c are compiled.
> 
> We're changing this to have conditional compilation of the .c files, now with a header just for this API, that will now error out if you _don't_ define something.

as previously mention, something will *always* be defined with our build method, this is mainly to prevent confusion for others compiling with alternative solutions

> I do think we should be able to specify the back-end with a define; no question; it's a great idea. I just think that it'd be simpler if instead of changing those other things, we just kept it in line with the other APIs, and the old compilation assumptions.

At some point this will end up bloating the dirmonitor.c

If we add a preprocessor check for every OS we are willing to support it will quickly add up and make the code unreadable.

Not to mention explicitly supporting all kqueue systems will be a nightmare.

> For example, if we just kept it all in one file, we could eliminate the header, and the glue defines, and basically just keep it as it was, but using an explicit define, rather than an OS one.

Jan200101 commented on src/api/dirmonitor.c:
```diff
@@ -15,135 +15,20 @@
+#include "dirmonitor.h"
 
 static int f_check_dir_callback(int watch_id, const char* path, void* L) {
   #if _WIN32
```
this needs to be updated to only be set when the WIN32 backend is enabled

Jan200101 commented on src/api/dirmonitor.h:
```diff
@@ -0,0 +1,27 @@
+*/
+struct dirmonitor {
+  int fd;
+  #if _WIN32
```
same here

Jan200101 commented on src/meson.build:
```diff
@@ -11,6 +10,37 @@ lite_sources = [
+    elif host_machine.system() == 'windows'
+        dirmonitor_backend = 'win32'
+    else
+        error('Could not find suitable dirmonitor backend')
```
generic posix backend could be added

adamharrison commented:
> this is practically the best way to do it
> sadly a lot of the dirmonitor code is dependent on the struct and there is no simple way of telling it it exists without informing it of its structure.

Right, but if you have it all in one file, it's just contained to the file; no need for a header.

> The header also allows us to cheat by setting defining the functions for a specific backend using preprocessors, which means any explit use of a certain backend will instantly cause compilation errors as well as define the function currectly so the compiler will not make assumptions

Right; but not having a header at all means there won't be any way to call a specific back-end outside the file; which I think is the reason why the APIs were set up the way they were (i.e. it's all self-contained).

> There is no easy way to do this with just preprocessors.
> We can't detect what functions we have or what headers we can include and, at best, can guess.
> There is the solution of defining kqueue and inotify_init in our code and dynamically resolving those at runtime, but I think that is way too overkill

Nono, I think you're good doing it in meson. I just mean use #if inside the file based on the backend define, rather than conditional compile *different* files.

> Currently in src/meson.build we detect the if the user is capable of using inotify, kqueue (either through native kqueue or libkqueue) or if they are on Windows, in which case we use the win32 backend.

Yes; this is good.

> The current system would not compile either unless at least one backend is present.
> It is trivial to add support for no backend or add a generic posix backend.

Could be good; we'd basically just need to return -1 for any of the calls; and the lua would handle it as-is, I think.

> I am very certain that this will mess up with build caching since it won't recompile dirmonitor.c unless it changed.

Agreed, I don't think we should do this.

> I am more fond of the DIRMONTOR_ build option solution, and it works well for projects like SDL2.
> Problem I see is that with meson we cannot programatically define options, so unless we are willing to let the meson_options.txt bloat I see this as unviable (also there is no #elif DIRMONITOR_KQUEUE`)

I agree, I think DIRMONITOR_ build option is a good call.

> as previously mention, something will always be defined with our build method, this is mainly to prevent confusion for others compiling with alternative solutions

Sure; but if you don't compile with meson, you can in theory not have the define. Again; I'm looking at it from an outsider's perspective, because I don't use meson. I basically see it as making way more sense to just keep it in one file; that way you can remove the source lines where you conditionally add sources in meson, and we can preserve the existing API paradigm across all APIs.

> At some point this will end up bloating the dirmonitor.c
> If we add a preprocessor check for every OS we are willing to support it will quickly add up and make the code unreadable.

Yeah; as I struckthrough; I don't think we need to do that. I think we just keep dirmonitor.c close to as-is, and replace the OS defines with the DIRMONITOR_ defines. Wouldn't that work? (and eliminate code?) I'm not at all against the DIRMONITOR_ defines, using meson to set them, and interpreting in the file; it's more just the file splitting and the conditional compilation.

Jan200101 commented:
Okay, this has spun into multiple different conversation points, I'll try to sum them up.

Right now the Code is separated using ifdefs depending on the platform.
The three sections are for win32, inotify and kqueue.

Objectively, this will become unreadable very quickly if we decide to add more backends, like FSEvents for MacOS or a generic Posix one.


The best solution for that is to separate all the backends in any way, shape or form so the code can stand independent of other platforms.

I think instead of continuing to put the code for 3 different backends into one function we should look how we should separate them instead.

adamharrison commented:
Right; I understand. I'm of the opposite opinion, for the reasons stated above; but that doesn't mean I'm right.

Let's wait for some other comments in one direction or another, and if we don't get any, we'll merge this as-is. Sound good?

jgmdev commented:
These 3 files already reside on a directory named `dirmonitor`, shouldn't the `dirmonitor_` prefix be removed?

* dirmonitor_inotify.c -> inotify.c
* dirmonitor_win32.c -> win32.c
* dirmonitor_kqueue.c -> kqueue.c

Jan200101 commented:
> These 3 files already reside on a directory named `dirmonitor`, shouldn't the `dirmonitor_` prefix be removed?
> 
>     * dirmonitor_inotify.c -> inotify.c
> 
>     * dirmonitor_win32.c -> win32.c
> 
>     * dirmonitor_kqueue.c -> kqueue.c

Sure, I see no problem with that.
I'll wait to see where this goes before doing anything though

Guldoman commented:
Is the header needed?
The function definitions could be moved in `dirmonitor.c` and the struct can be moved inside each implementation.

`dirmonitor.c` doesn't really need the definition of the struct as it's used as an opaque pointer.

Jan200101 commented:
> Is the header needed? The function definitions could be moved in `dirmonitor.c` and the struct can be moved inside each implementation.
> 
> `dirmonitor.c` doesn't really need the definition of the struct as it's used as an opaque pointer.

You are right, dirmonitor doesn't need the struct
but all the implementations do, and keeping multiple versions of the same struct would be redundant.

Jan200101 commented:
I moved the header containing the struct into the dirmonitor directory and moved all function declarations into the dirmonitor.c file.

I have also renamed all dirmonitor backend files to not have dirmonitor in their filename.

jgmdev commented:
> I moved the header containing the struct into the dirmonitor directory and moved all function declarations into the dirmonitor.c file.
> 
> I have also renamed all dirmonitor backend files to not have dirmonitor in their filename.

I think you didn't pushed the changes

Jan200101 commented:
> > I moved the header containing the struct into the dirmonitor directory and moved all function declarations into the dirmonitor.c file.
> > I have also renamed all dirmonitor backend files to not have dirmonitor in their filename.
> 
> I think you didn't pushed the changes

Pushing took a while, the Internet where I am is a bit unreliable

adamharrison commented:
> Is the header needed?
> The function definitions could be moved in dirmonitor.c and the struct can be moved inside each implementation.

> dirmonitor.c doesn't really need the definition of the struct as it's used as an opaque pointer.

I agree with @Guldoman here; if we're going to do this, let's just leave it as an opaque pointer; there's no need for a header. Just declare the struct differently in each .c file, instead of using #if (given that we're removing all of them anyway).

Jan200101 commented:
header has been replaced with an opaque pointer.

dummy backend added for targets that do not support inotify, kqueue or the win32 api.

added backend specific preprocessors so there can be conditional logic in the main dirmonitor.c file.

jgmdev commented:
Also just in case not to forget to include the latest changes Adam did to inotify here: https://github.com/lite-xl/lite-xl/pull/872/commits/cdeef6f23c3c87d861dfc3e8c3c745ff1aa17b6b

adamharrison commented:
Yeah; when I said that was the last thing you needed to do, turns out I lied. We do need to merge those changes from the commit. (Sorry).

Jan200101 commented:
MacOS fails because of general CI failure, PR unrelated

adamharrison commented:
One last thing I think we shoudl do is just set the default to be dummy if not specified, but this is fine, I suppose.
